File: pyimpl/tests/lexer_test.py
import pytest

from pyimpl import Token, TokenType, Lexer  # type: ignore


@pytest.mark.parametrize("literal, expected", [
    ("=", TokenType.ASSIGN),
    ("+", TokenType.PLUS),
    ("(", TokenType.LEFT_PAREN),
    (")", TokenType.RIGHT_PAREN),
    ("{", TokenType.LEFT_BRACE),
    ("}", TokenType.RIGHT_BRACE),
    (",", TokenType.COMMA),
    (";", TokenType.SEMICOLON),
    ("fn", TokenType.FUNCTION),
    ("let", TokenType.LET),
    ("x", TokenType.IDENTIFIER),
    ("int", TokenType.INT),
])
def test_token_emitted(literal: str, expected: TokenType) -> None:
    lexer = Lexer(literal)
    token = lexer.next_token()
    assert token.token_type == expected


def test_min_language_construct() -> None:
    input_tokens = ("   let five = 5;\n"
                    "   let ten = 10;\n"
                    "	let add = fn(x, y) { x + y };\n"
                    "	let result = add(five, ten);    ")

    expected_tokens = [
        # let five = 5;
        Token(TokenType.LET, "let"),
        Token(TokenType.IDENTIFIER, "five"),
        Token(TokenType.ASSIGN, "="),
        Token(TokenType.INT, "5"),
        Token(TokenType.SEMICOLON, ";"),

        # let ten = 10;
        Token(TokenType.LET, "let"),
        Token(TokenType.IDENTIFIER, "ten"),
        Token(TokenType.ASSIGN, "="),
        Token(TokenType.INT, "10"),
        Token(TokenType.SEMICOLON, ";"),

        # let add = fn(x, y) { x + y };
        Token(TokenType.LET, "let"),
        Token(TokenType.IDENTIFIER, "add"),
        Token(TokenType.ASSIGN, "="),
        Token(TokenType.FUNCTION, "fn"),
        Token(TokenType.LEFT_PAREN, "("),
        Token(TokenType.IDENTIFIER, "x"),
        Token(TokenType.COMMA, ","),
        Token(TokenType.IDENTIFIER, "y"),
        Token(TokenType.RIGHT_PAREN, ")"),
        Token(TokenType.LEFT_BRACE, "{"),
        Token(TokenType.IDENTIFIER, "x"),
        Token(TokenType.PLUS, "+"),
        Token(TokenType.IDENTIFIER, "y"),
        Token(TokenType.RIGHT_BRACE, "}"),
        Token(TokenType.SEMICOLON, ";"),

        # let result = add(five, ten);
        Token(TokenType.LET, "let"),
        Token(TokenType.IDENTIFIER, "result"),
        Token(TokenType.ASSIGN, "="),
        Token(TokenType.IDENTIFIER, "add"),
        Token(TokenType.LEFT_PAREN, "("),
        Token(TokenType.IDENTIFIER, "five"),
        Token(TokenType.COMMA, ","),
        Token(TokenType.IDENTIFIER, "ten"),
        Token(TokenType.RIGHT_PAREN, ")"),
        Token(TokenType.SEMICOLON, ";"),

        Token(TokenType.EOF, ""),
    ]

    assert lexable(input_tokens, expected_tokens)


def lexable(input_str: str, expected_tok: list[Token]) -> bool:
    lexer = Lexer(input_str)
    for index, expected_token in enumerate(expected_tok):
        token = lexer.next_token()
        if token != expected_token:
            print(f"\nexpected token: {expected_token} at index: {index}")
            print(f"actual token: {token}")
            return False
    return True

File: goimpl/lexer/lexer_test.go
package lexer

import (
	"goimpl/token"
	"testing"
)

func TestSimpleTokens(t *testing.T) {

	input := "=+,;(){}"

	tests := []expectedToken{
		{token.ASSIGN, "="},
		{token.PLUS, "+"},
		{token.COMMA, ","},
		{token.SEMICOLON, ";"},
		{token.LPRAN, "("},
		{token.RPRAN, ")"},
		{token.LBRACE, "{"},
		{token.RBRACE, "}"},
		{token.EOF, ""},
	}

	testLexer(input, tests, t)

}

func TestMinLanguageConstruct(t *testing.T) {

	input := `	let five = 5;
				let ten = 10;
				let mul = fn(x, y) { x * y };
				let add = fn(x, y) { x + y };
				let result = if(five < ten) { mul(five, ten) } else { add(five, ten) };
			`

	tests := []expectedToken{
		// let five = 5;
		{token.LET, "let"},
		{token.IDENTIFIER, "five"},
		{token.ASSIGN, "="},
		{token.INT, "5"},
		{token.SEMICOLON, ";"},

		// let ten = 10;
		{token.LET, "let"},
		{token.IDENTIFIER, "ten"},
		{token.ASSIGN, "="},
		{token.INT, "10"},
		{token.SEMICOLON, ";"},

		// let mul = fn(x, y) { x * y };
		{token.LET, "let"},
		{token.IDENTIFIER, "mul"},
		{token.ASSIGN, "="},
		{token.FUNCTION, "fn"},
		{token.LPRAN, "("},
		{token.IDENTIFIER, "x"},
		{token.COMMA, ","},
		{token.IDENTIFIER, "y"},
		{token.RPRAN, ")"},
		{token.LBRACE, "{"},
		{token.IDENTIFIER, "x"},
		{token.ASTER, "*"},
		{token.IDENTIFIER, "y"},
		{token.RBRACE, "}"},
		{token.SEMICOLON, ";"},

		// let add = fn(x, y) { x + y };
		{token.LET, "let"},
		{token.IDENTIFIER, "add"},
		{token.ASSIGN, "="},
		{token.FUNCTION, "fn"},
		{token.LPRAN, "("},
		{token.IDENTIFIER, "x"},
		{token.COMMA, ","},
		{token.IDENTIFIER, "y"},
		{token.RPRAN, ")"},
		{token.LBRACE, "{"},
		{token.IDENTIFIER, "x"},
		{token.PLUS, "+"},
		{token.IDENTIFIER, "y"},
		{token.RBRACE, "}"},
		{token.SEMICOLON, ";"},

		// let result = if(five < ten) { mul(five, ten) } else { add(five, ten) };
		{token.LET, "let"},
		{token.IDENTIFIER, "result"},
		{token.ASSIGN, "="},
		{token.IF, "if"},
		{token.LPRAN, "("},
		{token.IDENTIFIER, "five"},
		{token.LT, "<"},
		{token.IDENTIFIER, "ten"},
		{token.RPRAN, ")"},
		{token.LBRACE, "{"},
		{token.IDENTIFIER, "mul"},
		{token.LPRAN, "("},
		{token.IDENTIFIER, "five"},
		{token.COMMA, ","},
		{token.IDENTIFIER, "ten"},
		{token.RPRAN, ")"},
		{token.RBRACE, "}"},
		{token.ELSE, "else"},
		{token.LBRACE, "{"},
		{token.IDENTIFIER, "add"},
		{token.LPRAN, "("},
		{token.IDENTIFIER, "five"},
		{token.COMMA, ","},
		{token.IDENTIFIER, "ten"},
		{token.RPRAN, ")"},
		{token.RBRACE, "}"},
		{token.SEMICOLON, ";"},

		// EOF
		{token.EOF, ""},
	}

	testLexer(input, tests, t)

}

type expectedToken struct {
	expectedType    token.TokenType
	expectedLiteral string
}

func testLexer(input string, tests []expectedToken, t *testing.T) {
	l := NewLazyLexer(input)
	for i, tokTest := range tests {
		tok := l.NextToken()

		if tok.Type != tokTest.expectedType {
			got := tok.Type
			expected := tokTest.expectedType
			t.Fatalf("tests[%d] - tokentype wrong. expected=%q, got=%q", i, expected, got)
		}

		if tok.Literal != tokTest.expectedLiteral {
			got := tok.Literal
			expected := tokTest.expectedLiteral
			t.Fatalf("tests[%d] - literal wrong. expected=%q, got=%q", i, expected, got)
		}
	}
}
